### ISSUE TITLE
Leadership changes under heavy load cause hanging in state machine

### ISSUE THREAD

**Issue by rsm10** [etcd-io/etcd#8553]
Version: etcd raft. 3.1.5
High level problem: When operating a raft cluster under a lot of load; the election process gets stuck. Essentially when a leader changes; one of the nodes doesn't see this change. It then tries to send commits to the old leader, but the http Post hangs forever. 

So we have nodes A, B and C. A is the leader, but then it loses its leadership. B doesn't see this, and still Posts commits to A.

Dumped the goroutine traces on all sides, and traced the hang down to the following function in node.go; on the node that WAS the leader (node A). The "select" here hangs forever when using a context without a timeout; otherwise it always times out. And it is this hanging that causes Node B to hang on its http Post.

// Step advances the state machine using msgs. The ctx.Err() will be returned,
// if any.
func (n *node) step(ctx context.Context, m pb.Message) error {
	ch := n.recvc
	if m.Type == pb.MsgProp {
		ch = n.propc
	}

	select {
	case ch <- m:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	case <-n.done:
		return ErrStopped
	}
}

Node that gets stuck waiting for a POST to return (node B in my example). The line " raft.node: b60bed changed leader from b60bed to b406cb at term 39884" indicates this node *thinks* the leader has changed, but it seems like it never actually did from the logs on the other two nodes.

raft2017/09/13 14:49:02 INFO: b60bed [quorum:3] has received 3 MsgVoteResp votes and 0 vote rejections
raft2017/09/13 14:49:02 INFO: b60bed became leader at term 39883
raft2017/09/13 14:49:02 INFO: raft.node: b60bed elected leader b60bed at term 39883
raft2017/09/13 14:49:56 INFO: b60bed [logterm: 39883, index: 2329863, vote: b60bed] ignored MsgVote from b406cb [logterm: 39883, index: 2329863] at term 39883: lease is not expired (remaining ticks: 4)
raft2017/09/13 14:49:56 INFO: b60bed [term: 39883] received a MsgAppResp message with higher term from b406cb [term: 39884]
raft2017/09/13 14:49:56 INFO: b60bed became follower at term 39884
raft2017/09/13 14:49:56 INFO: raft.node: b60bed changed leader from b60bed to b406cb at term 39884

and then:

raft2017/09/13 14:52:18 INFO: b60bed [logterm: 39883, index: 2329863, vote: b3e575] ignored MsgVote from b406cb [logterm: 39883, index: 2329863] at term 39884: lease is not expired (remaining ticks: 8)
raft2017/09/13 14:52:20 INFO: b60bed [logterm: 39883, index: 2329863, vote: b3e575] ignored MsgVote from b3e575 [logterm: 39883, index: 2329863] at term 39884: lease is not expired (remaining ticks: 8)
raft2017/09/13 14:52:22 INFO: b60bed [logterm: 39883, index: 2329863, vote: b3e575] ignored MsgVote from b3e575 [logterm: 39883, index: 2329863] at term 39884: lease is not expired (remaining ticks: 8)
raft2017/09/13 14:52:24 INFO: b60bed [logterm: 39883, index: 2329863, vote: b3e575] ignored MsgVote from b3e575 [logterm: 39883, index: 2329863] at term 39884: lease is not expired (remaining ticks: 8)

Node keeps trying to be leader (Node A in my example), though stuck node (B) already thinks it is leader even though this node never received a Quorum of responses to become elected.

aft2017/09/13 14:49:02 INFO: raft.node: b406cb elected leader b60bed at term 39883
raft2017/09/13 14:49:55 INFO: b406cb is starting a new election at term 39883
raft2017/09/13 14:49:55 INFO: b406cb became candidate at term 39884
raft2017/09/13 14:49:55 INFO: b406cb received MsgVoteResp from b406cb at term 39884
raft2017/09/13 14:49:55 INFO: b406cb [logterm: 39883, index: 2329863] sent MsgVote request to ac720f at term 39884
raft2017/09/13 14:49:55 INFO: b406cb [logterm: 39883, index: 2329863] sent MsgVote request to b3e575 at term 39884
raft2017/09/13 14:49:55 INFO: b406cb [logterm: 39883, index: 2329863] sent MsgVote request to b60bed at term 39884
raft2017/09/13 14:49:55 INFO: raft.node: b406cb lost leader b60bed at term 39884
raft2017/09/13 14:49:59 INFO: b406cb is starting a new election at term 39884
raft2017/09/13 14:49:59 INFO: b406cb became candidate at term 39885
raft2017/09/13 14:49:59 INFO: b406cb received MsgVoteResp from b406cb at term 39885

Third node, triies to be leader; but by term 39886 it no longer tries. This node seems to be acting normally.

raft2017/09/13 14:49:02 INFO: b3e575 [term: 39882] received a MsgVote message with higher term from b60bed [term: 39883]
raft2017/09/13 14:49:02 INFO: b3e575 became follower at term 39883
raft2017/09/13 14:49:02 INFO: b3e575 [logterm: 39882, index: 2329385, vote: 0] cast MsgVote for b60bed [logterm: 39882, index: 2329385] at term 39883
raft2017/09/13 14:49:02 INFO: raft.node: b3e575 lost leader b60bed at term 39883
raft2017/09/13 14:49:02 INFO: raft.node: b3e575 elected leader b60bed at term 39883
raft2017/09/13 14:49:56 INFO: b3e575 [logterm: 39883, index: 2329863, vote: b60bed] ignored MsgVote from b406cb [logterm: 39883, index: 2329863] at term 39883: lease is not expired (remaining ticks: 10)
raft2017/09/13 14:49:59 INFO: b3e575 [logterm: 39883, index: 2329863, vote: b60bed] ignored MsgVote from b406cb [logterm: 39883, index: 2329863] at term 39883: lease is not expired (remaining ticks: 3)
raft2017/09/13 14:49:59 INFO: b3e575 is starting a new election at term 39883
raft2017/09/13 14:49:59 INFO: b3e575 became candidate at term 39884
raft2017/09/13 14:49:59 INFO: b3e575 received MsgVoteResp from b3e575 at term 39884
raft2017/09/13 14:49:59 INFO: b3e575 [logterm: 39883, index: 2329863] sent MsgVote request to ac720f at term 39884
raft2017/09/13 14:49:59 INFO: b3e575 [logterm: 39883, index: 2329863] sent MsgVote request to b406cb at term 39884
raft2017/09/13 14:49:59 INFO: b3e575 [logterm: 39883, index: 2329863] sent MsgVote request to b60bed at term 39884
raft2017/09/13 14:49:59 INFO: raft.node: b3e575 lost leader b60bed at term 39884
raft2017/09/13 14:50:02 INFO: b3e575 [term: 39884] received a MsgVote message with higher term from b406cb [term: 39886]
raft2017/09/13 14:50:02 INFO: b3e575 became follower at term 39886

**Comment by rsm10** [comment]
Note, it seems like there are two separate issues here.

1. The node that hangs when receiving messages meant for a leader; when it isn't the leader
2. The node that thinks the leader is elected even when it currently isn't and seemingly never was

**Comment by javaforfun** [comment]
I think this problem is fixed by https://github.com/coreos/etcd/pull/8339

**Comment by rsm10** [comment]
So 8339 definitely has helped, but I do still see POSTs timing out when a leader changes though in a different scenario:

ad05af was leader:
raft2017/09/14 19:43:50 INFO: ad05af became follower at term 260
raft2017/09/14 19:43:50 WARN: ad05af stepped down to follower since quorum is not active

On other two nodes, they get "stuck" again with the "lease is not expired" message from raft2017/09/14 19:43:54 to raft2017/09/14 19:44:56.

Other two nodes:


// It becomes a follower at term 284 when it receives a MsgVote with higher term; but not it does not automatically mark the sender as its leader. It then eventually settles on a new leader by term 287
raft2017/09/14 19:44:56 INFO: b560cc [term: 260] received a MsgApp message with higher term from ad05af [term: 284]
raft2017/09/14 19:44:56 INFO: b560cc became follower at term 284
raft2017/09/14 19:45:01 INFO: b560cc [term: 284] received a MsgVote message with higher term from ae7c41 [term: 285]
raft2017/09/14 19:45:01 INFO: b560cc became follower at term 285
raft2017/09/14 19:45:01 INFO: b560cc [logterm: 284, index: 3351388, vote: 0] rejected MsgVote from ae7c41 [logterm: 284, index: 3351382] at term 285
raft2017/09/14 19:45:01 INFO: raft.node: b560cc lost leader ad05af at term 285
raft2017/09/14 19:45:04 INFO: b560cc [term: 285] received a MsgVote message with higher term from ae7c41 [term: 286]
raft2017/09/14 19:45:04 INFO: b560cc became follower at term 286
raft2017/09/14 19:45:04 INFO: b560cc [logterm: 284, index: 3351388, vote: 0] rejected MsgVote from ae7c41 [logterm: 284, index: 3351382] at term 286
raft2017/09/14 19:45:06 INFO: b560cc is starting a new election at term 286
raft2017/09/14 19:45:06 INFO: b560cc became candidate at term 287
raft2017/09/14 19:45:06 INFO: b560cc received MsgVoteResp from b560cc at term 287
raft2017/09/14 19:45:06 INFO: b560cc [logterm: 284, index: 3351388] sent MsgVote request to ae7c41 at term 287
raft2017/09/14 19:45:06 INFO: b560cc [logterm: 284, index: 3351388] sent MsgVote request to ad05af at term 287
raft2017/09/14 19:45:06 INFO: b560cc received MsgVoteResp from ae7c41 at term 287
raft2017/09/14 19:45:06 INFO: b560cc [quorum:2] has received 2 MsgVoteResp votes and 0 vote rejections
raft2017/09/14 19:45:06 INFO: b560cc became leader at term 287
raft2017/09/14 19:45:06 INFO: raft.node: b560cc elected leader b560cc at term 287

// It becomes a follower at term 284 when it receives a MsgVote with higher term; but not it does not automatically mark the sender as its leader. It then receives the request from the original leader to be leader again; and votes YES at which point this node and the old leader (ad05af) both recognize ad05af as the current leader. Then more elections happen, and eventually b560cc becomes the leader at term 287.
raft2017/09/14 19:44:56 INFO: ae7c41 [term: 260] received a MsgVote message with higher term from ad05af [term: 284]
raft2017/09/14 19:44:56 INFO: ae7c41 became follower at term 284
raft2017/09/14 19:44:56 INFO: ae7c41 [logterm: 260, index: 3350555, vote: 0] cast MsgVote for ad05af [logterm: 260, index: 3351020] at term 284
raft2017/09/14 19:44:56 INFO: raft.node: ae7c41 lost leader ad05af at term 284
raft2017/09/14 19:44:56 INFO: raft.node: ae7c41 elected leader ad05af at term 284
raft2017/09/14 19:45:01 INFO: ae7c41 is starting a new election at term 284
raft2017/09/14 19:45:01 INFO: ae7c41 became candidate at term 285
raft2017/09/14 19:45:01 INFO: ae7c41 received MsgVoteResp from ae7c41 at term 285
raft2017/09/14 19:45:01 INFO: ae7c41 [logterm: 284, index: 3351382] sent MsgVote request to ad05af at term 285
raft2017/09/14 19:45:01 INFO: ae7c41 [logterm: 284, index: 3351382] sent MsgVote request to b560cc at term 285
raft2017/09/14 19:45:01 INFO: raft.node: ae7c41 lost leader ad05af at term 285
raft2017/09/14 19:45:01 INFO: ae7c41 received MsgVoteResp rejection from b560cc at term 285
raft2017/09/14 19:45:01 INFO: ae7c41 [quorum:2] has received 1 MsgVoteResp votes and 1 vote rejections
raft2017/09/14 19:45:04 INFO: ae7c41 is starting a new election at term 285
raft2017/09/14 19:45:04 INFO: ae7c41 became candidate at term 286
raft2017/09/14 19:45:04 INFO: ae7c41 received MsgVoteResp from ae7c41 at term 286
raft2017/09/14 19:45:04 INFO: ae7c41 [logterm: 284, index: 3351382] sent MsgVote request to ad05af at term 286
raft2017/09/14 19:45:04 INFO: ae7c41 [logterm: 284, index: 3351382] sent MsgVote request to b560cc at term 286
raft2017/09/14 19:45:04 INFO: ae7c41 received MsgVoteResp rejection from b560cc at term 286
raft2017/09/14 19:45:04 INFO: ae7c41 [quorum:2] has received 1 MsgVoteResp votes and 1 vote rejections
raft2017/09/14 19:45:06 INFO: ae7c41 [term: 286] received a MsgVote message with higher term from b560cc [term: 287]
raft2017/09/14 19:45:06 INFO: ae7c41 became follower at term 287
raft2017/09/14 19:45:06 INFO: ae7c41 [logterm: 284, index: 3351382, vote: 0] cast MsgVote for b560cc [logterm: 284, index: 3351388] at term 287
raft2017/09/14 19:45:06 INFO: raft.node: ae7c41 elected leader b560cc at term 287



And the old leader shows the following, which indicates at term 284 it thought it had become the leader but was not recognized by the others. It is briefly elected, but then overthrown by b560cc at term 287

raft2017/09/14 19:44:56 INFO: ad05af received MsgVoteResp from ae7c41 at term 284
raft2017/09/14 19:44:56 INFO: ad05af [quorum:2] has received 2 MsgVoteResp votes and 0 vote rejections
raft2017/09/14 19:44:56 INFO: ad05af became leader at term 284
raft2017/09/14 19:44:56 INFO: raft.node: ad05af elected leader ad05af at term 284
raft2017/09/14 19:45:01 INFO: ad05af [logterm: 284, index: 3351491, vote: ad05af] ignored MsgVote from ae7c41 [logterm: 284, index: 3351382] at term 284: lease is not expired (remaining ticks: 9)
raft2017/09/14 19:45:04 INFO: ad05af [logterm: 284, index: 3351586, vote: ad05af] ignored MsgVote from ae7c41 [logterm: 284, index: 3351382] at term 284: lease is not expired (remaining ticks: 9)
raft2017/09/14 19:45:04 INFO: ad05af [term: 284] received a MsgAppResp message with higher term from b560cc [term: 286]
raft2017/09/14 19:45:04 INFO: ad05af became follower at term 286
raft2017/09/14 19:45:04 INFO: raft.node: ad05af lost leader ad05af at term 286
raft2017/09/14 19:45:06 INFO: ad05af [term: 286] received a MsgVote message with higher term from b560cc [term: 287]
raft2017/09/14 19:45:06 INFO: ad05af became follower at term 287
raft2017/09/14 19:45:06 INFO: ad05af [logterm: 284, index: 3351597, vote: 0] rejected MsgVote from b560cc [logterm: 284, index: 3351388] at term 287
raft2017/09/14 19:45:06 INFO: found conflict at index 3351389 [existing term: 284, conflicting term: 287]
raft2017/09/14 19:45:06 INFO: replace the unstable entries from index 3351389
raft2017/09/14 19:45:06 INFO: raft.node: ad05af elected leader b560cc at term 287


From looking at the additional logging we have added to our cluster, it seems both non-leader nodes were creating snapshots at the time that the leader's checkQuorum returned false.

Non-leader nodes both had:

Sep 14 19:43:47 99-60 INFO: start snapshot [applied index: 3350544 | last snapshot index: 3300515]
Sep 14 19:43:53 99-60 INFO: Saving snapshot to disk...
Sep 14 19:43:54 99-60 INFO: compacted log at index 3345544

Which are at the roughly the same time span as:

raft2017/09/14 19:43:50 WARN: ad05af stepped down to follower since quorum is not active
raft2017/09/14 19:43:50 INFO: ad05af became follower at term 260
WARNING: HTTP status 500 while posting raft message to https://10.96.99.61:18003/v1/raft 

This seems to indicate this "since quorum is not active" which relates to checkQuorumActive in etcd/raft/raft.go is a direct result of these nodes being non-responsive while creating snapshots.

The "posting" error is because of a 30 second timeout we have on our http clients. This seems to indicate to me that there is still something where if a node (node A) sends a POST to a node it *thinks* is the leader (node B), but isn't elected yet; something on Node B doesn't correctly return an error immediately; and hence the client just infers an error via this timeout

**Comment by rsm10** [comment]
Confirmed I still see my original issue.  Still think it is related to this SELECT from node.go, that seems to only return because I set a context timeout. Something aboutt the channel used to wait on here is wrong when a message is received from a node that thinks this node is the leader, but this node does not think it is a leader.

// Step advances the state machine using msgs. The ctx.Err() will be returned,
// if any.
func (n *node) step(ctx context.Context, m pb.Message) error {
	ch := n.recvc
	if m.Type == pb.MsgProp {
		ch = n.propc
	}

	select {
	case ch <- m:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	case <-n.done:
		return ErrStopped
	}
}

**Comment by rsm10** [comment]
I am wondering if the "ch" channel here is nil; and hence sending to it blocks forever.

On a cluster I saw this issue, it seems like I can correlate a node losing leader status with these hanging posts.

On node that was receiving POSTs that hang:

raft2017/09/15 11:50:24 INFO: raft.node: aa82e9 lost leader aa82e9 at term 6606

On node that gets stuck:

raft2017/09/15 11:50:19 INFO: raft.node: b0f9c4 changed leader from ae49a7 to aa82e9 at term 6606
raft2017/09/15 11:50:28 INFO: b0f9c4 [logterm: 6606, index: 4104494, vote: 0] ignored MsgVote from aa82e9 [logterm: 6606, index: 4104494] at term 6606: lease is not expired (remaining ticks: 10)
raft2017/09/15 11:50:30 INFO: b0f9c4 [logterm: 6606, index: 4104494, vote: 0] ignored MsgVote from aa82e9 [logterm: 6606, index: 4104494] at term 6606: lease is not expired (remaining ticks: 10)
...
...


the lost leader log message comes from node.go, in "func (n *node) run(r *raft) {" and is associated with setting propc to nil. Perhaps this could be the root cause?

		if lead != r.lead {
			if r.hasLeader() {
				if lead == None {
					r.logger.Infof("raft.node: %x elected leader %x at term %d", r.id, r.lead, r.Term)
				} else {
					r.logger.Infof("raft.node: %x changed leader from %x to %x at term %d", r.id, lead, r.lead, r.Term)
				}
				propc = n.propc
			} else {
				r.logger.Infof("raft.node: %x lost leader %x at term %d", r.id, lead, r.Term)
				propc = nil
			}
			lead = r.lead
		}

**Comment by rsm10** [comment]
so added some debugging to a patched version of this code, and appears that in "func (n *node) step(ctx context.Context, m pb.Message) error {"; the 'ch' channel is NOT nil after all.

So maybe it is full?

**Comment by xiang90** [comment]
@rsm10 This issue is not super helpful. Can you:

1. format your logging lines to make it easier to read
2. tell us more about your context/backgroud
3. tell us how to reproduce it

**Comment by rsm10** [comment]
I'll need to try and reproduce in a simpler way if you want something that you can take, run and repro with. At a high level, I am just running a raft cluster with a very rapid, continuous load of commits. We are using raft essentially as a replicated queue.

I don't see this issue all the time. Generally only after several hours of hammering my raft cluster with data to be committed does it pop up.  The symptoms are that I start to see my "node.node.Step" calls timeout (because I use a context with a 30 second timeout). I have dumped the goroutines using some pprof calls, and can say with certainty that node.node.Step hangs on those lines I've already mentioned Also, the node that is hanging on node.node.Step happens to think there is no leader; while other nodes incorrectly think there is a leader (seems normal for this to be a temporary condition when leaders change).

node.go, line 43, which is the select in this function:

func (n *node) step(ctx context.Context, m pb.Message) error {
	ch := n.recvc
	if m.Type == pb.MsgProp {
		ch = n.propc
	}

	select {
	case ch <- m:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	case <-n.done:
		return ErrStopped
	}
}

I added some debug (as I mentioned) so I know the 'ch' channel is not nil. The only other way I know that this select would hang is if the channel 'ch' points to is full and nothing is pulling from it.

Looking at this logic some more, the node.go run function (func (n *node) run(r *raft) {) has logic which assigns to a propc var based on some conditionals.
The only way n.propc would be read from (hence unblocking the send to it above, assuming ch does indeed point to n.propc; would be if r.hasLeader returns true (which it shouldn't in this case).

		if lead != r.lead {
			if r.hasLeader() {
				if lead == None {
					r.logger.Infof("raft.node: %x elected leader %x at term %d", r.id, r.lead, r.Term)
				} else {
					r.logger.Infof("raft.node: %x changed leader from %x to %x at term %d", r.id, lead, r.lead, r.Term)
				}
				propc = n.propc
			} else {
				r.logger.Infof("raft.node: %x lost leader %x at term %d", r.id, lead, r.Term)
				propc = nil
			}
			lead = r.lead
		}

So basically when this node receives messages from other nodes that think it is the leader; but it actually isnt; it seems to hang until the context times it out. In my previous comments I've already correlated the " lost leader" log message with this issue. Which supports my theory of there being a sort of channel sender/receiver mismatch here. node.run() is blocked receiving FROM n.recvc but node.step is blocked sending TO n.propc; and hence the sender blocks forever.

I can format my logging lines better, but without giving you my exact raft cluster code (which I can't); I don't know how to explain in any more detail how to reproduce this.

**Comment by rsm10** [comment]
if the problem is as I describe, what about something like this as a speculative solution:
in node.go. run function

change this:
		select {
		// TODO: maybe buffer the config propose if there exists one (the way
		// described in raft dissertation)
		// Currently it is dropped in Step silently.
		case m := <-propc:
			m.From = r.id
			r.Step(m)

to:
		case m := <-n.propc:
			if lead != r.lead && r.hasLeader() {
				m.From = r.id
				r.Step(m)
			}

this at least would unblock that sender and would basically ignore messages that this node doesn't expect to handle

**Comment by javaforfun** [comment]
```
Non-leader nodes both had:

Sep 14 19:43:47 99-60 INFO: start snapshot [applied index: 3350544 | last snapshot index: 3300515]
Sep 14 19:43:53 99-60 INFO: Saving snapshot to disk...
Sep 14 19:43:54 99-60 INFO: compacted log at index 3345544

Which are at the roughly the same time span as:

raft2017/09/14 19:43:50 WARN: ad05af stepped down to follower since quorum is not active
raft2017/09/14 19:43:50 INFO: ad05af became follower at term 260
WARNING: HTTP status 500 while posting raft message to https://10.96.99.61:18003/v1/raft
```
@rsm10 seems you only use one thread for raft and snapshot.  raft leader hang when creating snapshot (can't send Heartbeat, MsgApp...), so followers thinks leader timeout, then start election.
are you testing this?  https://github.com/coreos/etcd/blob/master/contrib/raftexample/raft.go#L439

**Comment by rsm10** [comment]
We are doing something similar to that example yes. That also has the snapshot and main raft loop on a single thread too no?


On Sep 15, 2017 8:31 PM, Shuai Li <notifications@github.com> wrote:

Non-leader nodes both had:

Sep 14 19:43:47 99-60 INFO: start snapshot [applied index: 3350544 | last snapshot index: 3300515]
Sep 14 19:43:53 99-60 INFO: Saving snapshot to disk...
Sep 14 19:43:54 99-60 INFO: compacted log at index 3345544

Which are at the roughly the same time span as:

raft2017/09/14 19:43:50 WARN: ad05af stepped down to follower since quorum is not active
raft2017/09/14 19:43:50 INFO: ad05af became follower at term 260
WARNING: HTTP status 500 while posting raft message to https://10.96.99.61:18003/v1/raft


@rsm10<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frsm10&data=02%7C01%7Crobin_mahony%40hotmail.com%7C46fb18d3f65e419df16808d4fcb35f0f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636411294691916975&sdata=ZzUQan9G7uUOMciEWc5KkgeQ50HUfBg69FhnchL%2BXyk%3D&reserved=0> seems you only use one thread for raft and snapshot. raft leader hang when creating snapshot (can't send Heartbeat, MsgApp...), so followers thinks leader timeout, then start election.
are you testing this? https://github.com/coreos/etcd/blob/master/contrib/raftexample/raft.go#L439<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fcoreos%2Fetcd%2Fblob%2Fmaster%2Fcontrib%2Fraftexample%2Fraft.go%23L439&data=02%7C01%7Crobin_mahony%40hotmail.com%7C46fb18d3f65e419df16808d4fcb35f0f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636411294691916975&sdata=gq8%2FG6bIdSjd3CPCL8PHca1IA55q9k3zgwdR%2F243Iyk%3D&reserved=0>

**Comment by xiang90** [comment]
@rsm10 have you figured out the root cause? it is because of the blocking snapshot?

**Comment by rsm10** [comment]
Definitely think the blocking snapshot is one of the causes. Working on moving this off the main raft loop. 
Though still think that it seems incorrect to me that there is a code path that can result in a blocking send; that only exits if a Context with timeout is used (as per my description earlier).

**Comment by xiang90** [comment]
any update on this issue?

**Comment by rsm10** [comment]
Had some issues with moving the snapshots off the main raft loop, so still working on that. Though my previous comment about how that other behaviour still appears to be a bug and how this isn't fixing that issue but more just trying to minimize its occurrence still stands.


________________________________

**Comment by sumithreddy** [comment]
> if the problem is as I describe, what about something like this as a speculative solution: in node.go. run function
> 
> change this: select { // TODO: maybe buffer the config propose if there exists one (the way // described in raft dissertation) // Currently it is dropped in Step silently. case m := <-propc: m.From = r.id r.Step(m)
> 
> to: case m := <-n.propc: if lead != r.lead && r.hasLeader() { m.From = r.id r.Step(m) }
> 
> this at least would unblock that sender and would basically ignore messages that this node doesn't expect to handle

@robin865 how did you handle this issue without raft lib fix. I am trying to pass in a context with timeout to the raft step but doesn't completely solve the issue.
